File: src/main/java/net/swade/permissionss/group/GroupManager.java
package net.swade.permissionss.group;

import cn.nukkit.Player;
import cn.nukkit.utils.Config;
import cn.nukkit.utils.ConfigSection;
import net.swade.permissionss.Main;
import net.swade.permissionss.Profile;
import net.swade.permissionss.Utils;
import net.swade.permissionss.enums.Process;
import net.swade.permissionss.permission.PermissionManager;

import java.util.*;

public class GroupManager {

    public static List<Group> groups;

    public static void load() {
        groups = new ArrayList<>();
        Config config = new Config(Main.getInstance().getDataFolder() + "/groups.yml", 2);
        Map<String, Object> map = config.getAll();
        map.forEach((string, object) -> {
            ConfigSection section = config.getSection(string);
            Group group = new Group(section.getString("name"),
                    section.getString("id"),
                    section.getString("nameTagFormat"),
                    section.getString("chatFormat"),
                    section.getStringList("aliases"),
                    section.getStringList("permissions"),
                    section.getStringList("inheritance"));
            groups.add(group);
        });
    }

    public static Group getGroup(String string) {
        for (Group value : groups) {
            if (value.getName().equalsIgnoreCase(string) || value.getAliases().contains(string.toLowerCase()) || value.getId().equalsIgnoreCase(string) ) {
                return value;
            }
        }

        return null;
    }

    public static Group getDefaultGroup(){
        Config config = Main.getInstance().getConfig();
        return getGroup(config.getString("defaultGroup"));
    }

    public static void setDefaultGroup(Group group){
        Config config = Main.getInstance().getConfig();
        config.set("defaultGroup", group.getId());
        config.save();
    }

    public static Group getPlayerGroup(Player player){
        return getPlayerGroup(player.getName());
    }

    public static Group getPlayerGroup(String name){
        Config config = new Config(Main.getInstance().getPlayerConfigPath(), 1);
        if (config.exists(name.toLowerCase())){
            Profile profile = Profile.getProfile(name);
            return getGroup(profile.getGroup());
        }
        Main.getInstance().registerPlayer(name);
        return getDefaultGroup();
    }

    public static void setPlayerGroup(Player player, Group group){
        setPlayerGroup(player.getName(), group);
    }

    public static void setPlayerGroup(String name, Group group){
        Config config = new Config(Main.getInstance().getPlayerConfigPath(), 1);
        Profile profile = Profile.getProfile(name);
        profile.setGroup(group.getId());
        config.set(name.toLowerCase(), profile.toString());
        config.save();
        PermissionManager.reloadPermissions();
    }

    public static Process addGroup(String groupName){
        if (!Utils.isValidGroupName(groupName)){
            return Process.INVALID_NAME;
        }else if (getGroup(groupName) != null){
            return Process.ALREADY_EXISTS;
        }

        Config config = new Config(Main.getInstance().getDataFolder() + "/groups.yml", 2);
        config.set(groupName +".name", groupName);
        config.set(groupName +".id", groupName.toLowerCase());
        config.set(groupName +".nameTagFormat", "§7[§a"+ groupName +"§7] %nickname%");
        config.set(groupName +".chatFormat", "§7[§a"+ groupName +"§7] %nickname% §e» §f%message%");
        config.set(groupName +".aliases", new ArrayList<>());
        config.set(groupName +".permissions", new ArrayList<>());
        config.set(groupName +".inheritance", new ArrayList<>());
        config.save();
        load();
        return Process.SUCCESS;
    }

    public static Process removeGroup(String groupName){
        if (!Utils.isValidGroupName(groupName)){
            return Process.INVALID_NAME;
        }else if (getGroup(groupName) == null){
            return Process.NOT_FOUND;
        }
        Group group = getGroup(groupName);
        Config config = new Config(Main.getInstance().getDataFolder() + "/groups.yml", 2);
        config.remove(group.getName());
        config.save();
        load();
        return Process.SUCCESS;
    }

    public static void setChatFormat(Group group, String format){
        Config config = new Config(Main.getInstance().getDataFolder() + "/groups.yml", 2);
        config.set(group.getId() +".chatFormat", format);
        config.save();
        load();
    }

    public static void setNameTagFormat(Group group, String format){
        Config config = new Config(Main.getInstance().getDataFolder() + "/groups.yml", 2);
        config.set(group.getId() +".nameTagFormat", format);
        config.save();
        load();
    }
}

File: src/main/kotlin/group/GroupManager.kt
package net.guneyilmaz0.permissions.group

import cn.nukkit.Player
import cn.nukkit.Server
import cn.nukkit.utils.Config
import cn.nukkit.utils.ConfigSection
import net.guneyilmaz0.permissions.PermissionsS
import net.guneyilmaz0.permissions.Utils
import net.guneyilmaz0.permissions.enums.Process
import net.guneyilmaz0.permissions.permission.PermissionManager
import java.util.UUID

object GroupManager {

    var groups: MutableList<Group> = mutableListOf()

    fun load() {
        groups.clear()
        val config = Config("${PermissionsS.instance.dataFolder}/groups.yml", Config.YAML)
        val map: Map<String, Any> = config.all
        map.forEach { (string, _) ->
            val section: ConfigSection = config.getSection(string)
            val group = Group(
                section.getString("name")!!,
                section.getString("id")!!,
                section.getString("nameTagFormat")!!,
                section.getString("chatFormat")!!,
                section.getStringList("aliases"),
                section.getStringList("permissions"),
                section.getStringList("inheritance")
            )
            groups.add(group)
        }
    }

    fun getGroup(name: String): Group? {
        return groups.find {
            it.name.equals(name, ignoreCase = true) ||
                    it.aliases.contains(name.lowercase()) ||
                    it.id.equals(name, ignoreCase = true)
        }
    }

    fun getGroupIds(): List<String> = groups.map { it.id }


    fun getDefaultGroup(): Group {
        val config = PermissionsS.instance.config
        return getGroup(config.getString("defaultGroup")) ?: run {
            val group = groups.first()
            setDefaultGroup(group)
            group
        }
    }

    fun setDefaultGroup(group: Group) {
        val config = PermissionsS.instance.config
        config.set("defaultGroup", group.id)
        config.save()
    }

    fun getPlayerGroup(player: Player): Group {
        if (PermissionsS.provider.isProfileExists(player.uniqueId))
            return getGroup(PermissionsS.provider.getProfile(player.uniqueId)!!.group) ?: getDefaultGroup()
        else {
            PermissionsS.provider.registerPlayer(player.uniqueId, player.name)
            return getDefaultGroup()
        }
    }

    fun setPlayerGroup(player: Player, group: Group) {
        setPlayerGroup(player.uniqueId, group)
    }

    fun setPlayerGroup(uuid: UUID, group: Group) {
        if (!PermissionsS.provider.isProfileExists(uuid)) {
            throw IllegalArgumentException("Profile not found")
        }
        val profile = PermissionsS.provider.getProfile(uuid)!!
        profile.group = group.id
        PermissionsS.provider.saveProfile(profile)
        val player = Server.getInstance().getOfflinePlayer(uuid)
        if (player.isOnline) PermissionManager.handlePermissions(player.player!!)
    }

    fun addGroup(groupName: String): Process {
        return when {
            Utils.isInvalidGroupName(groupName) -> Process.INVALID_NAME
            getGroup(groupName) != null -> Process.ALREADY_EXISTS
            else -> {
                val config = Config("${PermissionsS.instance.dataFolder}/groups.yml", Config.YAML)
                config.set("$groupName.name", groupName)
                config.set("$groupName.id", groupName.lowercase())
                config.set("$groupName.nameTagFormat", "§7[§a$groupName§7] %nickname%")
                config.set("$groupName.chatFormat", "§7[§a$groupName§7] %nickname% §e» §f%message%")
                config.set("$groupName.aliases", mutableListOf<String>())
                config.set("$groupName.permissions", mutableListOf<String>())
                config.set("$groupName.inheritance", mutableListOf<String>())
                config.save()
                load()
                Process.SUCCESS
            }
        }
    }

    fun removeGroup(groupName: String): Process {
        return when {
            Utils.isInvalidGroupName(groupName) -> Process.INVALID_NAME
            getGroup(groupName) == null -> Process.NOT_FOUND
            else -> {
                val group = getGroup(groupName) ?: return Process.NOT_FOUND
                val config = Config("${PermissionsS.instance.dataFolder}/groups.yml", Config.YAML)
                config.remove(group.name)
                config.save()
                load()
                Process.SUCCESS
            }
        }
    }

    fun setChatFormat(group: Group, format: String) {
        val config = Config("${PermissionsS.instance.dataFolder}/groups.yml", Config.YAML)
        config.set("${group.id}.chatFormat", format)
        config.save()
        load()
    }
}
